gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,267 @@
+"""Interface class from GtsfmData to PatchmatchNetData
+
+Authors: Ren Liu
+"""
+from typing import Any, Dict, List, Tuple
+
+import cv2
+import numpy as np
+from torch.utils.data import Dataset
+
+from gtsfm.common.gtsfm_data import GtsfmData
+from gtsfm.common.image import Image
+from gtsfm.densify.mvs_math import piecewise_gaussian
+
+NUM_PATCHMATCHNET_STAGES = 4
+
+MIN_DEPTH_PERCENTILE = 1
+MAX_DEPTH_PERCENTILE = 99
+
+
+class PatchmatchNetData(Dataset):
+    """PatchmatchNetData class for PatchmatchNet. It contains the interface from GtsfmData.
+    Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/datasets/dtu_yao_eval.py
+    """
+
+    def __init__(self, images: Dict[int, Image], sfm_result: GtsfmData, num_views: int = 5) -> None:
+        """Cache image and camera pose metadata for PatchmatchNet inference.
+
+        Args:
+            images: input images (H, W, C) to GTSFM
+            sfm_result: sfm results calculated by GTSFM
+            num_views: number of views, containing 1 reference view and (num_views-1) source views
+        """
+        assert images is not None and len(images) > 1
+
+        # Cache sfm result
+        self._sfm_result = sfm_result
+
+        # PatchmatchNet meta
+        self._num_views = num_views
+        self._num_stages = NUM_PATCHMATCHNET_STAGES
+
+        # Test data preparation
+        self._patchmatchnet_idx_to_camera_idx = sorted(self._sfm_result.get_valid_camera_indices())
+        self._num_images = len(self._patchmatchnet_idx_to_camera_idx)
+
+        # Store locations of camera centers in the world frame and create mapping from image indices (with some entries
+        #   potentially missing) to [0,N-1] tensor indices for N inputs to PatchmatchNet
+        self._camera_centers = {}
+        self._camera_idx_to_patchmatchnet_idx = {}
+        for i in range(self._num_images):
+            self._camera_centers[self._patchmatchnet_idx_to_camera_idx[i]] = (
+                self._sfm_result.get_camera(self._patchmatchnet_idx_to_camera_idx[i]).pose().translation()
+            )
+            self._camera_idx_to_patchmatchnet_idx[self._patchmatchnet_idx_to_camera_idx[i]] = i
+
+        self._images = images
+
+        # Set image dimensions to those of zero’th image
+        self._h, self._w = (
+            self._images[self._patchmatchnet_idx_to_camera_idx[0]].height,
+            self._images[self._patchmatchnet_idx_to_camera_idx[0]].width,
+        )
+        # Calculate the cropped size of each image so that the height and width of the image can be divided by 8 evenly,
+        #   so that the image size can be matched after downsampling and then upsampling in PatchmatchNet
+        self._cropped_h, self._cropped_w = (self._h - self._h % 8, self._w - self._w % 8)
+
+        self._pairs, self._depth_ranges = self.configure()
+
+    def configure(self) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
there is a lot of complex code here, so I think we have to have a unit test on configure. could we make a simple scenario, with 3 cameras around a circle, with two or three 3d points, projected into each camera?